File: src-tauri/src/commands/projects.rs
use tauri::State;

use crate::{
    models::{
        errors::AppResult,
        project::{CreateProjectRequest, Project, RenameProjectRequest},
    },
    AppState,
};

#[tauri::command]
pub fn list_projects(state: State<'_, AppState>) -> AppResult<Vec<Project>> {
    state.db.list_projects()
}

#[tauri::command]
pub fn create_project(
    state: State<'_, AppState>,
    request: CreateProjectRequest,
) -> AppResult<Project> {
    state.db.create_project(&request.name)
}

#[tauri::command]
pub fn rename_project(
    state: State<'_, AppState>,
    id: String,
    request: RenameProjectRequest,
) -> AppResult<Project> {
    state.db.rename_project(&id, &request.name)
}

#[tauri::command]
pub fn delete_project(state: State<'_, AppState>, id: String) -> AppResult<()> {
    state.db.delete_project(&id)
}

#[tauri::command]
pub fn assign_generation_to_project(
    state: State<'_, AppState>,
    generation_id: String,
    project_id: Option<String>,
) -> AppResult<()> {
    state
        .db
        .set_generation_project(&generation_id, project_id.as_deref())
}

File: src/app/lib/store/slices/projects.ts
import type { GenerationStore } from "@/app/lib/store/types";
import type { StoreApi } from "zustand";
import * as api from "@/app/lib/api";

export function createProjectsSlice(
  set: StoreApi<GenerationStore>["setState"],
  _get: StoreApi<GenerationStore>["getState"],
) {
  return {
    projects: [] as GenerationStore["projects"],
    activeProjectId: null,

    refreshProjects: async () => {
      if (!api.isTauriRuntime()) return;
      const projects = await api.listProjects();
      set({ projects });
    },

    createProject: async (name: string) => {
      if (!api.isTauriRuntime()) return;
      const project = await api.createProject(name);
      set((state) => ({ projects: [project, ...state.projects] }));
    },

    renameProject: async (id: string, name: string) => {
      if (!api.isTauriRuntime()) return;
      const updated = await api.renameProject(id, name);
      set((state) => ({
        projects: state.projects.map((p) => (p.id === id ? updated : p)),
      }));
    },

    deleteProject: async (id: string) => {
      if (!api.isTauriRuntime()) return;
      await api.deleteProject(id);
      set((state) => ({
        projects: state.projects.filter((p) => p.id !== id),
        activeProjectId: state.activeProjectId === id ? null : state.activeProjectId,
        history: state.history.map((r) => (r.projectId === id ? { ...r, projectId: null } : r)),
      }));
    },

    setActiveProject: (id: string | null) => {
      set({ activeProjectId: id });
    },

    assignGenerationToProject: async (generationId: string, projectId: string | null) => {
      if (!api.isTauriRuntime()) return;
      await api.assignGenerationToProject(generationId, projectId);
      set((state) => ({
        history: state.history.map((r) => (r.id === generationId ? { ...r, projectId } : r)),
      }));
    },
  };
}
